### ISSUE TITLE
Make kl_div accept not just `target`, but `log_target` as well to avoid numerical issues when `target=softmax`, for example.

### ISSUE THREAD

**Issue by nikitaved** [pytorch/pytorch#34536]
## 🚀 Feature
Make `kl_div` accept `target` in both normal and log-space to avoid potential numerical issues.

## Motivation

The current implementation computes `log(target)` explicitly, which could be sub-optimal numerically. Some probabilities might have some structure which could be used to increase the stability of `log(target)`, for example in `softmax`, where instead of computing the log of ratios, it is preferred to compute the difference of logs. 

## Pitch

Eliminate the explicit `log(target)` computation in the `kl_div`, but allow it to accept `log_target` as an additional optional parameter. So, if `log_target == None`, then the implementation resorts to explicit but potentially less numerically stable `log(target)`.

## Additional context
See [#32520](https://github.com/pytorch/pytorch/issues/32520) for a round example.

CC @ezyang @rgommers @v0dro @gchanan

**Comment by gchanan** [comment]
we'd accept a flag that indicates if the target is in the log space or not.

**Comment by rgommers** [comment]
Done in commit 35cdb785228b8a, closing.
